src/sage/rings/polynomial/laurent_polynomial_mpair.pyx

```diff
@@ -1172,17 +1172,68 @@ cdef class LaurentPolynomial_mpair(LaurentPolynomial):
         if not x:
             return self._poly.total_degree() + sum(self._mon)
 
+        # Get the index of the generator or error
         cdef tuple g = <tuple > self._parent.gens()
         cdef Py_ssize_t i
-        cdef bint no_generator_found = True
-        for i in range(len(g)):
-            if g[i] is x:
-                no_generator_found = False
-                break
-        if no_generator_found:
-            raise TypeError("x must be a generator of parent")
+        try:
+            i = g.index(x)
+        except ValueError:  # not in the tuple
+            raise TypeError(f"{x} is not a generator of parent")
         return self._poly.degree(self._parent._R.gens()[i]) + self._mon[i]
 
+    def valuation(self, x=None):
+        """
+        Return the valuation of ``self``.
+
+        INPUT:
+
+        - ``x`` -- (optional) a generator; if given, return the valuation
+          with respect to this generator
+
+        EXAMPLES::
+
+            sage: R.<x,y> = LaurentPolynomialRing(ZZ)
+            sage: f = 2*x^2*y^-3 - 13*x^-1*y^-3 + 2*x^2*y^-5 - 2*x^-3*y^2
+            sage: f.valuation()
+            -4
+            sage: f.valuation(x)
+            -3
+            sage: f.valuation(y)
+            -5
+            sage: R.zero().valuation()
+            +Infinity
+
+        TESTS:
+
+        If supplied, ``x`` must be a generator::
+
+            sage: R.<x,y> = LaurentPolynomialRing(ZZ)
+            sage: f = 1 + x + x^2*y^-1
+            sage: f.valuation(1)
+            Traceback (most recent call last):
+            ...
+            TypeError: 1 is not a generator of parent
+        """
+        # Valuation of zero polynomial is defined to be +Infinity
+        if self.is_zero():
+            return Infinity
+
+        # When x is None find the minimal valuation by finding the minimal
+        # valuation of the sum of exponents
+        if x is None:
+            return Integer(min(sum(e) for e in self.exponents()))
+
+        # Get the index of the generator or error
+        cdef tuple g = <tuple > self._parent.gens()
```

Comment:
this makes me think that an index of a generator would also be valid input. It's much more direct.

Argument against that would be that in principle, any irreducible polynomial could be valid input, since it would describe a prime divisor at which a valuation can be computed. Whether that needs to be implemented is another matter.

Reply:
Hmm this is a good point. I think if we do this for `valuation()` then we must also do this for `degree()` which has an almost identical form.

I'm happy to do the change which allows `x` to be either a generator or index. Does anyone else have an opinion?